_posts/2021-01-13-hapi-fhir.md

```diff
@@ -0,0 +1,454 @@
+The following code snippets comes from the HAPI FHIR documentation
+
+
+#### CORS
+
+The HAPI FHIR server framework includes an interceptor that can be used to provide CORS functionality on your server. 
+This mechanism relies purely on Java configuration. 
+HAPI's interceptor is a thin wrapper around Spring Framework's CorsProcessor class, 
+so it requires Spring to be present on your classpath.
+
+  
+
+```Java
+
+@WebServlet(urlPatterns = {"/fhir/*"}, displayName = "FHIR Server")
+public class RestfulServerWithCors extends RestfulServer {
+    @Override
+    protected  void  initialize() throws  ServletException {
+    // ... define your resource providers here ...
+    // Define your CORS configuration. This is an example
+    // showing a typical setup. You should customize this
+    // to your specific needs
+    CorsConfiguration  config = new  CorsConfiguration();
+    config.addAllowedHeader("x-fhir-starter");
+    config.addAllowedHeader("Origin");
+    config.addAllowedHeader("Accept");
+    config.addAllowedHeader("X-Requested-With");
+    config.addAllowedHeader("Content-Type");
+    config.addAllowedOrigin("*");
+    config.addExposedHeader("Location");
+    config.addExposedHeader("Content-Location");
+    config.setAllowedMethods(Arrays.asList("GET", "POST", "PUT", "DELETE", "OPTIONS", "PATCH"));
+    // Create the interceptor and register it
+    CorsInterceptor  interceptor = new  CorsInterceptor(config);
+    registerInterceptor(interceptor);
+    }
+}
+```
+
+#### Authorization Interceptor
+
+HAPI FHIR provides you with an Authorization interceptor which can be helpful with the task of determining whether a user has the appropriate permission to perform a given task on a FHIR server. 
+This is done by declaring a set of rules that can selectively allow (whitelist) and/or selectively block (blacklist) requests. 
+The interceptor works by allowing you to declare permission based on an individual request coming in.
+
+  
+
+The AuthorizationInterceptor is used by subclassing it and then registering your subclass with the RestfulServer. 
+The following example shows a subclassed interceptor implementing some basic rules:
+
+```java
+@SuppressWarnings("ConstantConditions")
+public class PatientAndAdminAuthorizationInterceptor extends AuthorizationInterceptor {
+
+    @Override
+    public  List<IAuthRule> buildRuleList(RequestDetails  theRequestDetails) {
+        // Process authorization header - The following is a fake
+        // implementation. Obviously we'd want something more real
+        // for a production scenario.
+        //
+        // In this basic example we have two hardcoded bearer tokens,
+        // one which is for a user that has access to one patient, and
+        // another that has full access.
+        IdType  userIdPatientId = null;
+        boolean  userIsAdmin = false;
+        String  authHeader = theRequestDetails.getHeader("Authorization");
+        if ("Bearer dfw98h38r".equals(authHeader)) {
```

Comment:
Better to  create 2 private methods: isNormalUser(authHeader) and isAdmin(authHeader) which check the bearer token